llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
       DL);
 }
 
+static SDValue
+PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
+  auto VT = N->getValueType(0);
+  if (!DCI.isAfterLegalizeDAG() || !Isv2x16VT(VT))
+    return SDValue();
+
+  auto Op0 = N->getOperand(0);
+  auto Op1 = N->getOperand(1);
+
+  // Start out by assuming we want to take the lower 2 bytes of each i32
+  // operand.
+  uint64_t Op0Bytes = 0x10;
+  uint64_t Op1Bytes = 0x54;
+
+  std::pair<SDValue *, uint64_t *> OpData[2] = {{&Op0, &Op0Bytes},
+                                                {&Op1, &Op1Bytes}};
+
+  // Check that each operand is an i16, truncated from an i32 operand. We'll
+  // select individual bytes from those original operands. Optionally, fold in a
+  // shift right of that original operand.
+  for (auto &[Op, OpBytes] : OpData) {
+    // Eat up any bitcast
+    if (Op->getOpcode() == ISD::BITCAST)
+      *Op = Op->getOperand(0);
+
+    if (Op->getValueType() != MVT::i16 || Op->getOpcode() != ISD::TRUNCATE ||
```

Comment:
It might be good to check that some of these instructions only have a single use, otherwise this transform is not a clear win and may increase register pressure.

Reply:
I'll write some tests to check what happens. Artem in another comment thinks that it might be okay if the instructions aren't one-use.

Reply:
I've tried some simple tests with multiple uses of the truncate and/or the original value.

When you reuse the truncate you appear to increase register pressure (though the SASS remains the same): https://godbolt.org/z/MT8PqshsW

When you reuse the original value the register pressure looks better, indicating the PRMT is worthwhile. Though the SASS is the same: https://godbolt.org/z/MqdbT6W59

When you reuse both, the register pressure is still worse, though the SASS remains the same: https://godbolt.org/z/Mc8W763M8

So, even though the SASS remains the same in these simple examples, it indicates we should probably bail out if the truncate has multiple uses. Multiple uses of the original value appears to be alright.

Reply:
Oh yes I should add, those links are using ptxas v11.5.0. It's the same result with the most recent 12.6.1; the SASS is always the same between the PRMT and non-PRMT versions.

Reply:
I've pushed a change that forbids multiple uses of the truncate, and have added tests for it.

Reply:
> the register pressure is still worse, though the SASS remains the same

What register pressure are you referring to here? Is it the number of declared registers: 
```
.reg .b32       %r<number of registers declared>;
```

---

Do we know of a case where the SASS differs due to this change?

Reply:
Yeah, I was referring to the number declared registers. I thought that's what you were referring to as it's the only thing we have "control" of.

I've only tried the examples I linked to above but no I've not seen a difference in the actual SASS. I could imagine that more PTX registers could manifest in worse SASS in more complex examples? It's not much to go on, though.

Reply:
On the tests we're consistently declaring four less `b16` registers, and one more `.b32`. It looks like a net win to me, even if it does not matter for SASS. It means we're doing less unnecessary stuff with `.b16` intermediate values.